⚠️ change catalog-specific URL from full path to API endpoint ref (#429)
operator-framework/catalogd: docs/fetching-catalog-contents.md

# `ClusterCatalog` Interface

> `catalogd` serves catalog content via a catalog-specific, versioned HTTP(S) endpoint. Clients access catalog information via this API endpoint and a versioned reference of the desired format. Current support includes only a complete catalog download, indicated by the path "v1/all", for example `https://catalogd-catalogservice.olmv1-system.svc/catalogs/operatorhubio/api/vi/all` would receive the complete FBC for the catalog `operatorhubio`.

Reviewer: ```suggestion
`catalogd` serves catalog content via a catalog-specific, versioned HTTP(S) endpoint. Clients access catalog information via this API endpoint and a versioned reference of the desired format. Current support includes only a complete catalog download, indicated by the path "v1/all", for example `https://catalogd-catalogservice.olmv1-system.svc/catalogs/operatorhubio/api/v1/all` would receive the complete FBC for the catalog `operatorhubio`.
```

Reviewer: I also think we should say something like `if status.baseURL is "https://base", then "https://base/v1/all" would receive the complete FBC...`

Or something like that to clarify that the baseURL itself is still opaque and that clients are expected to join the documentated API paths with the baseURL from the `ClusterCatalog` status.

> `catalogd` serves catalog content via a catalog-specific, versioned HTTP(S) endpoint. Clients access catalog information via this API endpoint and a versioned reference of the desired format. Current support includes only a complete catalog download, indicated by the path "v1/all", for example if `status.urls.base` is `https://catalogd-service.olmv1-system.svc/catalogs/operatorhub/api` then `https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/vi/all` would receive the complete FBC for the catalog `operatorhubio`.

Reviewer 2: Ideally it should be https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

> `catalogd` serves catalog content via a catalog-specific, versioned HTTP(S) endpoint. Clients access catalog information via this API endpoint and a versioned reference of the desired format. Current support includes only a complete catalog download, indicated by the path "v1/all", for example if `status.urls.base` is `https://catalogd-service.olmv1-system.svc/catalogs/operatorhub/api` then `https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/vi/all` would receive the complete FBC for the catalog `operatorhubio`.

Reviewer 2: Also not a blocker for merging this PR.

> `catalogd` serves catalog content via a catalog-specific, versioned HTTP(S) endpoint. Clients access catalog information via this API endpoint and a versioned reference of the desired format. Current support includes only a complete catalog download, indicated by the path "v1/all", for example if `status.urls.base` is `https://catalogd-service.olmv1-system.svc/catalogs/operatorhub/api` then `https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/vi/all` would receive the complete FBC for the catalog `operatorhubio`.

Reviewer: ```suggestion
`catalogd` serves catalog content via a catalog-specific, versioned HTTP(S) endpoint. Clients access catalog information via this API endpoint and a versioned reference of the desired format. Current support includes only a complete catalog download, indicated by the path "/api/v1/all", for example if `status.urls.base` is `https://catalogd-service.olmv1-system.svc/catalogs/operatorhub` then `https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/vi/all` would receive the complete FBC for the catalog `operatorhubio`.
```

Author: `status.urls[].base` for catalogs includes the `/api` base path: 
```sh
k get clustercatalog operatorhubio -o yaml | yq '.status.urls[]'
https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api
```

> `catalogd` serves catalog content via a catalog-specific, versioned HTTP(S) endpoint. Clients access catalog information via this API endpoint and a versioned reference of the desired format. Current support includes only a complete catalog download, indicated by the path "v1/all", for example if `status.urls.base` is `https://catalogd-service.olmv1-system.svc/catalogs/operatorhub/api` then `https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/vi/all` would receive the complete FBC for the catalog `operatorhubio`.

Author: That one is a side-effect of the fact that the previous implementation is hosting a filesystem via an http handler, and returns 404 for anything which is not explicitly found in that filesystem.  As we evolve this I imagine we will move away from this approach and implement something which can be this nuanced.

For local development, consider skipping TLS verification, such as `curl -k`, or reference external material
on self-signed certificate verification.

`ClusterCatalog` CRs have a status.baseURL field which identifies the catalog-specific API to access the catalog content:

```yaml
  status:
    .
    .

>     baseURL: https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/

Reviewer: ```suggestion
    urls:
        base: https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio
```

>     baseURL: https://catalogd-catalogservice.olmv1-system.svc/catalogs/operatorhubio/api/

Reviewer: We replaced `catalogservice` with just `service`, right?

```suggestion
    baseURL: https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/
```

    resolvedSource:
      image:
        ref: quay.io/operatorhubio/catalog@sha256:e53267559addc85227c2a7901ca54b980bc900276fc24d3f4db0549cb38ecf76
      type: Image
```

## On cluster

When making a request for the complete contents of the `operatorhubio` `ClusterCatalog` from within

> the cluster, clients would combine `status.baseURL` with the desired API service and issue an HTTP GET request for the URL.

Reviewer: ```suggestion
the cluster, clients would combine `status.urls.base` with the desired API service and issue an HTTP GET request for the URL.
```

For example, to receive the complete catalog data for the `operatorhubio` catalog indicated above, the client would append the service point designator `v1/all`, like:

> `https://catalogd-catalogservice.olmv1-system.svc/catalogs/operatorhubio/api/v1/all`.

Reviewer: ```suggestion
`https://catalogd-service.olmv1-system.svc/catalogs/operatorhubio/api/v1/all`.
```
